nixos/modules/hardware/device-tree.nix

```diff
@@ -121,8 +212,25 @@ in
           example = literalExpression "pkgs.linux_latest";
           type = types.path;
           description = lib.mdDoc ''
-            Kernel package containing the base device-tree (.dtb) to boot. Uses
-            device trees bundled with the Linux kernel by default.
+            Kernel package where device tree include directory is from. Also used as default source of dtb package to apply overlays to
+          '';
+        };
+
+        dtsSource = mkOption {
+          default = null;
+          type = types.nullOr dtsSourceType;
+          description = lib.mdDoc ''
+            Dts source that is used to compile into .dtb that serves as base to apply overlay to. This will take precedence over the {option}`hardware.deviceTree.dtbSource` option if both are specified.
+          '';
+        };
+
+        dtbSource = mkOption {
+          default = "${cfg.kernelPackage}/dtbs";
+          defaultText = literalExpression "\${cfg.kernelPackage}/dtbs";
+          type = types.path;
+          description = lib.mdDoc ''
+            Path to dtb directory that overlays and other processing will be applied to. Uses
+            device trees bundled with the Linux kernel by default. Will be ignored if {option}`hardware.deviceTree.dtsSource` is also specified.
```

Comment:
Config options should _compose_ rather than conflict with each other. It should be possible for a user to combine all of these options in a useful way.

For example, suppose I want to produce an image supporting four boards:

- Board 1 has a device tree in the Linux kernel. I want to use the DTB from the compiled `kernelPackage`.
- Board 2 needs a compiled `dtb` from the vendor. I want to include that with `dtbSource`.
- Board 3 has my own `dts` that I need to build. I would use `dtsSource` for this.
- Board 4 also comes from my own `dts`. I want to specify `dtsSource` multiple times.
- Then, I might want to apply a `filter` on top to build images tailored to each board.

As a user, I would expect this to "just work". Each option should build upon others. Something like this:

- multiple `dtsSource` -> custom `dtbs` tree
- custom `dtbs` tree + user-supplied `dtbSource` + `${kernelPackage}/dtbs` -> combined `dtb` tree
- combined `dtb` tree + `filter` -> filtered `dtb` tree
- filtered `dtb` tree + `overlays` -> final `dtb` tree

All of these configs could have sensible defaults that can be overridden. But that override should be an _explicit choice_ by the user and preferably not automatic behavior behind the scenes, even if it is documented.

For example, the default value of `dtbSources` could include both `kernelPackages` dtbs and dtbs compiled from `dtsSource`. If a user doesn't want to include dtbs from `kernelPackages`, they could set `kernelPackages = null`.